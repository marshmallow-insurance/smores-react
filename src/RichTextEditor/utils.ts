import { $createParagraphNode, LexicalNode, RootNode } from 'lexical'

export const appendNodes = (root: RootNode, nodes: LexicalNode[]) => {
  nodes
    .filter((node) => node.__type !== 'linebreak')
    .map((node) => {
      if (node.__type === 'text') {
        const paragraphNode = $createParagraphNode()
        paragraphNode.append(node)
        return paragraphNode
      }
      return node
    })
    .forEach((node) => root.append(node))
}
