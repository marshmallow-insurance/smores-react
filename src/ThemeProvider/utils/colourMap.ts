import * as designTokens from '@mrshmllw/smores-foundations/build/web/variables.json'
import { Color } from 'theme'

type Flatten<T, Prefix extends string = ''> = {
  [K in keyof T & string]: T[K] extends Record<string, any>
    ? Flatten<T[K], `${Prefix}${K}.`>
    : `${Prefix}${K}`
}[keyof T & string]

export type NewColor = Flatten<
  Pick<typeof designTokens, 'color' | 'extended1' | 'thirdParty'>
>

export type ColorTypes = Color | NewColor

// TODO: Test how well this works
// we currently access numerical colours like `theme.color.neutral[100]` or in the case of 000 colours, `theme.color.neutral['000']`
export const getThemeColor = (path: NewColor): string => {
  return path
    .split('.')
    .reduce((acc, key) => acc?.[key], designTokens as any) as string
}

// a function that returns a flattened dot notation string path to access the color value
export const formatDesignTokenColor = (): string => {
  const { color, extended1, thirdParty } = designTokens
  const colors = { color, extended1, thirdParty }

  if (colors === null) {
    return ''
  }

  const result: string[] = []
  const destructureNestedObject = (obj: any, currentPath: string) => {
    for (const key in obj) {
      if (Object.prototype.hasOwnProperty.call(obj, key)) {
        const value = obj[key]
        const newPath = currentPath ? `${currentPath}.${key}` : key
        if (typeof value === 'object' && value !== null) {
          destructureNestedObject(value, newPath)
        } else {
          result.push(newPath)
        }
      }
    }
  }
  destructureNestedObject(colors, '')
  return result.join(', ')
}

// old colour name â†’ new colour path // base token value
export const legacyColorMap: Record<Color, NewColor> = {
  fairyFloss: 'color.surface.brand.100', // palette.primary.040
  bubblegum: 'color.surface.brand.200', // palette.primary.060
  marshmallowPink: 'color.surface.brand.300', // palette.primary.100
  lollipop: 'color.surface.brand.400', // palatte.primary.120

  // Core Secondary
  chia: 'color.feedback.inactive.100', // palette.secondary.040
  sesame: 'color.text.subtle', // palette.secondary.060
  liquorice: 'color.text.base', // palette.secondary.100
  boba: 'color.text.contrast', // palette.secondary.120

  // Core Tertiary
  cream: `color.surface.base.000`, // palette.neutral.000
  coconut: 'color.surface.base.100', // palette.neutral.010
  mascarpone: 'color.surface.base.200', // palette.neutral.020
  custard: 'color.surface.base.300', // palette.neutral.040

  // Brand Secondary
  feijoa: 'color.illustration.accent1.100', // palette.brand1.060
  spearmint: 'color.illustration.accent1.200', // palette.brand1.100
  macaroon: 'color.illustration.accent2.100', // palette.brand2.060
  blueberry: 'color.illustration.accent2.200', // palette.brand2.100
  pistachio: 'color.illustration.accent3.200', // palette.brand3.100
  matcha: 'color.illustration.accent3.100', // palette.brand3.060
  caramel: 'color.illustration.accent4.200', // palette.brand4.100
  peanut: 'color.illustration.accent4.100', // palette.brand4.060
  marzipan: 'color.illustration.neutral.400', // palette.neutral.100
  oatmeal: 'color.illustration.neutral.300', // palette.neutral.060
  satsuma: 'extended1.20', // extended1.020

  // Traffic light
  watermelon: 'color.feedback.negative.100', // palette.negative.020
  strawberry: 'color.feedback.negative.200', // palette.negative.100
  apple: 'color.feedback.positive.200', // palette.positive.100
  mint: 'color.feedback.positive.100', // palette.positive.020
  lemon: 'color.feedback.notice.200', // palette.notice.100
  sherbert: 'color.feedback.notice.100', // palette.notice.020
  tangerine: 'extended1.100', // extended1.100

  // Third-party brand colours
  compareTheMarket: 'thirdParty.compareTheMarket',
  confused: 'thirdParty.confusedCom',
  onfido: 'thirdParty.onfido',
  x: 'thirdParty.twitter', // x rebrand not yet reflected in foundations
  premfina: 'thirdParty.premfina',
  checkout: 'thirdParty.checkout',
  meta: 'thirdParty.facebook', // Meta rebrand not yet reflected in foundations
  stripe: 'thirdParty.stripe',
  intercom: 'thirdParty.intercom',
  ravelin: 'thirdParty.ravelin',
  rac: 'thirdParty.rac',
  hometree: 'thirdParty.hometree',
} as const

/* usage example */
// prop change:
// change color prop type to be `color?: Color | NewColor`
// component usage:
// <Component color={'marshmallowPink'}/>
// OR
// <Component color={'interactive.primary.base'}/>
// inside component:
// const resolvedColor = color in legacyColorMap ? getThemeColor(legacyColorMap[color]) : getThemeColor(color)

// the above should resolve to #FF88C8
