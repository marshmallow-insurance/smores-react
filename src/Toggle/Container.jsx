import React, {useState} from 'react';
import {Toggle} from './Toggle';

export const Container = () => {
  const [active, setActive] = useState(false);

  return <Toggle checked={active} onToggle={() => setActive(!active)} />;
};
