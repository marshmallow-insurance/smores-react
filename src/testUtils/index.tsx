import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { SmoresThemeProvider } from '../ThemeProvider/ThemeProvider'

const Providers = ({ children }: { children: React.ReactNode }) => {
  return <SmoresThemeProvider>{children}</SmoresThemeProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react'
export * from '@testing-library/jest-dom/vitest'
export * from 'jest-styled-components'

export { customRender as render }
