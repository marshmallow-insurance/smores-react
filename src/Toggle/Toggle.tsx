import React, { FC } from 'react'
import styled from 'styled-components'

import { focusOutline } from '../utils/focusOutline'
import { theme } from '../theme'

type Props = {
  /** unique ID */
  id?: string
  /** checked flag  */
  checked: boolean
  /** onToggle listener  */
  onToggle: () => void
}

export const Toggle: FC<Props> = ({ checked, onToggle }) => {
  return (
    <Switch>
      <Checkbox type="checkbox" checked={checked} onChange={onToggle} />
      <Slider />
    </Switch>
  )
}

const Switch = styled.label`
  position: relative;
  display: inline-block;
  width: 50px;
  height: 32px;

  input {
    opacity: 0;
    width: 0;
    height: 0;
  }
`

const Slider = styled.span`
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: ${theme.colors.bgPrimary};
  border: 1px solid;
  border-color: ${theme.colors.bgPrimary};
  border-radius: 28px;
  transition: 0.2s background-color border-color;
  outline: none;

  &:before {
    position: absolute;
    content: '';
    height: 24px;
    width: 24px;
    left: 3px;
    bottom: 3px;
    background-color: ${theme.colors.white};
    transition: 0.2s transform;
    border-radius: 50%;
  }
`

const Checkbox = styled.input`
  ${focusOutline({ selector: `&:focus-visible + ${Slider}` })}

  &:checked + ${Slider} {
    background-color: ${theme.colors.success};
    border: 1px solid;
    border-color: ${theme.colors.success};
  }

  &:checked + ${Slider}:before {
    transform: translateX(18px);
  }
`
