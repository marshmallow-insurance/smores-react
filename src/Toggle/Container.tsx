import React, { useState } from 'react'
import { Toggle } from './Toggle'

export const Container = () => {
  const [active, setActive] = useState(false)

  return (
    <Toggle
      id="toggleId"
      checked={active}
      onToggle={() => setActive(!active)}
    />
  )
}
