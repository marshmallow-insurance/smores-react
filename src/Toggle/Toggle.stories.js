import React from 'react'
import { storiesOf } from '@storybook/react'

import { Container } from './Container.jsx'

storiesOf('Toggle', module).addWithJSX('default', () => <Container />)
