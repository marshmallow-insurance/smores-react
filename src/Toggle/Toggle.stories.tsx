import React from 'react'
import { Toggle } from './Toggle'
import { Container } from './Container'

export default {
  title: 'Toggle',
  component: Toggle,
}

const Template = () => <Container />

export const Default = Template.bind({})
