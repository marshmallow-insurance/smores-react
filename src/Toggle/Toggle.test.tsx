import React from 'react'
import { render } from '../testUtils'
import { it, expect } from 'vitest'
import { Toggle } from './Toggle'
import { noop } from '../utils/noop'

describe('Toggle Component', () => {
  it('renders correctly with default props', () => {
    const { container } = render(
      <Toggle
        checked={true}
        onToggle={noop}
        bgColor="color.illustration.accent4.100"
      />,
    )
    expect(container).toMatchSnapshot()
  })
})
