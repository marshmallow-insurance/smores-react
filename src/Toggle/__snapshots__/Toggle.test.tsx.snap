// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toggle Component > renders correctly with default props 1`] = `
<div>
  <label
    class="sc-bRKDuR bHjpxP sc-hvigdm eJKeDG"
  >
    <input
      aria-label="toggle"
      checked=""
      class="sc-ggWZvA lkdDSc"
      type="checkbox"
    />
    <span
      class="sc-fhHczv iAVPse"
    />
  </label>
</div>
`;
